### PR TITLE
gnome-extensions-cli: init at 0.9.5

### LINKED ISSUE #222099: Package request: gnome-extensions-cli
**Project description**

Allows you to manage GNOME extensions from CLI

**Metadata**

* homepage URL: https://pypi.org/project/gnome-extensions-cli/
* source URL: https://github.com/essembeh/gnome-extensions-cli
* license: apache 2
* platforms: linux

### PR DESCRIPTION
**Pull request by dylanmtaylor**
###### Description of changes

Add a package for gnome-extensions-cli. This closes #222099.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by dylanmtaylor** [comment]
Builds and runs as expected.

image

File: pkgs/desktops/gnome/misc/gnome-extensions-cli/default.nix
Comment (jtojnar) [inline pkgs/desktops/gnome/misc/gnome-extensions-cli/default.nix:8]: Should not this use `buildPythonApplication`?

File: pkgs/desktops/gnome/misc/gnome-extensions-cli/default.nix
Comment (jtojnar) [inline pkgs/desktops/gnome/misc/gnome-extensions-cli/default.nix:40]: GNOME should be capitalized.

File: pkgs/desktops/gnome/misc/gnome-extensions-cli/default.nix
Comment (dylanmtaylor) [reply pkgs/desktops/gnome/misc/gnome-extensions-cli/default.nix:8]: I'm not sure, I can change it if this is preferred.

File: pkgs/desktops/gnome/misc/gnome-extensions-cli/default.nix
Comment (dylanmtaylor) [reply pkgs/desktops/gnome/misc/gnome-extensions-cli/default.nix:40]: Consider it done. I'll edit it with the buildPythonApplication change.

**Comment by dylanmtaylor** [comment]
Just did a quick test, it still builds and runs with it as Application instead of Package.

**Comment by dylanmtaylor** [comment]
@jtojnar I have force pushed with the requested changes. Could you please re-review this? Thanks!

**Comment by dylanmtaylor** [comment]
Rebased onto master.

File: pkgs/desktops/gnome/misc/gnome-extensions-cli/default.nix
Comment (dylanmtaylor) [reply pkgs/desktops/gnome/misc/gnome-extensions-cli/default.nix]: @jtojnar I changed this in the latest commit. Looks like either would have worked, but buildPythonApplication is more appropriate. https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/python.section.md#buildpythonapplication-function-buildpythonapplication-function

Can you review this PR, please?

**Comment by dylanmtaylor** [comment]
I just realized that you must be extremely busy with the GNOME 44 work, @jtojnar. I saw the gnome branch and it looks like an incredible amount of work - this PR can understandably wait until that's all done :)

File: pkgs/desktops/gnome/misc/gnome-extensions-cli/default.nix
Comment (lilyinstarlight) [inline pkgs/desktops/gnome/misc/gnome-extensions-cli/default.nix:21]: Do you think you could get the upstream tests to run using `pytestCheckHook` in `nativeBuildInputs`?

**Review by lilyinstarlight — COMMENTED** [review]
This looks mostly good aside from the `with` usage! I've left a few comments to address

File: pkgs/desktops/gnome/misc/gnome-extensions-cli/default.nix
Comment (dylanmtaylor) [reply pkgs/desktops/gnome/misc/gnome-extensions-cli/default.nix:19]: Possibly but I don't know if that will work with CI since it looks like the tests require it to interact with a GNOME environment on the local machine via dbus. If one isn't up and running I'd imagine it would fail.

Comment (dylanmtaylor) [reply pkgs/desktops/gnome/misc/gnome-extensions-cli/default.nix:19]: https://github.com/essembeh/gnome-extensions-cli/blob/main/tests/test_dbus.py

Comment (dylanmtaylor) [reply pkgs/desktops/gnome/misc/gnome-extensions-cli/default.nix:19]: Actually, I see that it has a skip if not available. Maybe I can get that working.

@pytest.mark.skipif(not test_dbus_available(True), reason="DBus is not available")

Comment (dylanmtaylor) [reply pkgs/desktops/gnome/misc/gnome-extensions-cli/default.nix:19]: Hmm. That is failing.

Executing pytestCheckPhase
============================= test session starts ==============================
platform linux -- Python 3.10.10, pytest-7.2.1, pluggy-1.0.0
rootdir: /build/gnome_extensions_cli-0.9.5
collected 0 items                                                              

============================ no tests ran in 0.01s =============================
/nix/store/pw17yc3mwmsci4jygwalj8ppg0drz31v-stdenv-linux/setup: line 1593: pop_var_context: head of shell_variables not a function context
error: builder for '/nix/store/95g2b3lqh6w76ybp0lsswpy3i1lharn0-gnome-extensions-cli-0.9.5.drv' failed with exit code 5;
       last 10 log lines:
       > Check whether the following modules can be imported: gnome_extensions_cli
       > pytestCheckPhase
       > Executing pytestCheckPhase
       > ============================= test session starts ==============================
       > platform linux -- Python 3.10.10, pytest-7.2.1, pluggy-1.0.0
       > rootdir: /build/gnome_extensions_cli-0.9.5
       > collected 0 items                                                              
       >
       > ============================ no tests ran in 0.01s =============================
       > /nix/store/pw17yc3mwmsci4jygwalj8ppg0drz31v-stdenv-linux/setup: line 1593: pop_var_context: head of shell_variables not a function context
       For full logs, run 'nix log /nix/store/95g2b3lqh6w76ybp0lsswpy3i1lharn0-gnome-extensions-cli-0.9.5.drv'.

Comment (dylanmtaylor) [reply pkgs/desktops/gnome/misc/gnome-extensions-cli/default.nix:19]: I'm not sure how to fix this one myself, but I'm open to accepting help on it.

Comment (lilyinstarlight) [reply pkgs/desktops/gnome/misc/gnome-extensions-cli/default.nix:19]: Oh it's probably because you are fetching from PyPI instead of from GitHub. PyPI tarballs are often missing tests if they are separate from the package folders. If this doesn't get tests, it's not that big of a deal though (just always nice to have of course)

File: pkgs/desktops/gnome/misc/gnome-extensions-cli/default.nix
Comment (lilyinstarlight) [inline pkgs/desktops/gnome/misc/gnome-extensions-cli/default.nix:6]: Since you used `python3Packages.callPackage` in `all-packages.nix`, you should be able to accept `buildPythonApplication`, `poetry-core`, `colorama`, etc all as arguments at the top of the file

Was there a particular reason to only accept the `python3` argument?

Comment (dylanmtaylor) [reply pkgs/desktops/gnome/misc/gnome-extensions-cli/default.nix:6]: There was not, I suppose I did it that way due to inexperience. I can change that.

File: pkgs/desktops/gnome/misc/gnome-extensions-cli/default.nix
Comment (dylanmtaylor) [reply pkgs/desktops/gnome/misc/gnome-extensions-cli/default.nix:19]: This is a good call, I looked at the tarball contents and it indeed doesn't have the tests in it.

File: pkgs/desktops/gnome/misc/gnome-extensions-cli/default.nix
Comment (lilyinstarlight) [reply pkgs/desktops/gnome/misc/gnome-extensions-cli/default.nix:6]: No worries! :)

I was just making sure there wasn't a reason for this before more strongly suggesting a change

Comment (dylanmtaylor) [reply pkgs/desktops/gnome/misc/gnome-extensions-cli/default.nix:6]: Let me go ahead and move those to arguments and drop with python3.pkgs.

Comment (dylanmtaylor) [reply pkgs/desktops/gnome/misc/gnome-extensions-cli/default.nix]: I just pushed. Is this what you were thinking?

**Comment by dylanmtaylor** [comment]
Dropped typer in latest commit.

File: pkgs/desktops/gnome/misc/gnome-extensions-cli/default.nix
Comment (lilyinstarlight) [inline pkgs/desktops/gnome/misc/gnome-extensions-cli/default.nix:2]: Okay, one last nit and this looks good to me

```suggestion
```

This arg isn't used anymore

Comment (dylanmtaylor) [reply pkgs/desktops/gnome/misc/gnome-extensions-cli/default.nix]: Fixed :)

**Review by lilyinstarlight — APPROVED** [review]
This looks good to me now!

Did you have further input @jtojnar, since this is under the GNOME umbrella and you're code-owner?

**Review by jtojnar — APPROVED** [review]
Have not tested but the code looks alright to me (though not sure how useful it is when many extensions installed from the portal will not have the necessary patches).

**Comment by dylanmtaylor** [comment]
> Have not tested but the code looks alright to me (though not sure how useful it is when many extensions installed from the portal will not have the necessary patches).

I personally use this in conjunction with home-manager to manage extensions, and it seems to work fine. I think this is good to be merged in now @lilyinstarlight

**Comment by lilyinstarlight** [comment]
So I actually decided to try and test it, but it seems trying to use the dbus backend fails with `ModuleNotFoundError: No module named 'gi'` and trying to use the filesystem backend expects `gnome-shell` and `gsettings` in PATH

Perhaps this needs `pygobject3` in `propagatedBuildInputs`? That seems to take care of most of that

**Comment by dylanmtaylor** [comment]
> So I actually decided to try and test it, but it seems trying to use the dbus backend fails with `ModuleNotFoundError: No module named 'gi'` and trying to use the filesystem backend expects `gnome-shell` and `gsettingsAre ` in PATH
> 
> Perhaps this needs `pygobject3` in `propagatedBuildInputs`? That seems to take care of most of that

You're on GNOME? This is odd, I had it working in this comment. Perhaps I just happened to have the right dependency installed by chance? https://github.com/NixOS/nixpkgs/pull/222254#issuecomment-1477166398

**Comment by dylanmtaylor** [comment]
> pygobject3

Let me add it.

**Comment by jtojnar** [comment]
If this is GI-based, you will need to add `gobject-introspection` and `wrapGAppsNoGuiHook` to `nativeBuildInputs`. Maybe you ran it from a Python-based Terminal emulator if it did not complain about missing `gi`? You can try to run it with empty `XDG_DATA_DIRS` environment variable to detect missing GI typelib dependencies.

**Comment by dylanmtaylor** [comment]
I am pretty sure that sceenshot of when I tested it was from GNOME console on NixOS. Super weird. Let me try to fix it, I'll spin up a fresh untainted NixOS VM and try it there so I have a clean environment.

**Comment by lilyinstarlight** [comment]
> Maybe you ran it from a Python-based Terminal emulator if it did not complain about missing `gi`?

It seems to only need GI-based stuff when the DBus backend is selected. By default it tries to import it but then falls back to a non-GI based backend when that fails. Which is why it didn't complain

**Comment by lilyinstarlight** [comment]
Looking at the source code, it only uses Gio from gobject and that's just to handle DBus

**Comment by dylanmtaylor** [comment]
> Looking at the source code, it only uses Gio from gobject and that's just to handle DBus

So you think that the pygobject3 will fix this? Still waiting on my virtual machine to install.

**Comment by lilyinstarlight** [comment]
> So you think that the pygobject3 will fix this? Still waiting on my virtual machine to install.

@jtojnar knows *way* more about it than I do, so I would defer to what they say. I only mentioned that because adding it locally seemed to be enough to get everything to work (I didn't entirely clear my environment either, just `PATH`, so the whole `wrapGAppsNoGuiHook` might be needed)

**Comment by dylanmtaylor** [comment]
> > So you think that the pygobject3 will fix this? Still waiting on my virtual machine to install.
> 
> @jtojnar knows _way_ more about it than I do, so I would defer to what they say. I only mentioned that because adding it locally seemed to be enough to get everything to work (I didn't entirely clear my environment either, just `PATH`, so the whole `wrapGAppsNoGuiHook` might be needed)

Makes sense. I'll try building it in the VM and see if it's functional - I want to get this right so that it works as expected on a new NixOS install.  I am hopeful I'll get it in before 23.05 is released, but there's plenty of time for that :)

**Comment by jtojnar** [comment]
I think gobject-introspection makes the glib libraries like `Gio` available automatically. But it is probably best if `glib` is in `GI_TYPELIB_PATH` added by the wrapper just to be sure.

**Comment by dylanmtaylor** [comment]
This is on a fresh NixOS install:
image
I saved the latest default.nix to my home directory, ran`nix-build -E 'with import <nixpkgs> {}; recurseIntoAttrs (python3Packages.callPackage ./default.nix {})'`
and then cd'd to result/bin and ran ./gext install dash-to-panel@jderose9.github.com

It looks like this works. I am happy to add the gobject-introspection though.

**Comment by dylanmtaylor** [comment]
> it is probably best if `glib` is in `GI_TYPELIB_PATH` added by the wrapper just to be sure.

It isn't clear from this comment what you think I need to edit to make this work. Can you clarify this bit?

**Comment by dylanmtaylor** [comment]
Latest push adds gobject-introspection as a nativeBuildInput. 

> > it is probably best if `glib` is in `GI_TYPELIB_PATH` added by the wrapper just to be sure.
> 
> It isn't clear from this comment what you think I need to edit to make this work. Can you clarify this bit?

Oh, I get it now, this is just saying I need to add wrapGAppsNoGuiHook. Let me add that too just in case.

**Comment by dylanmtaylor** [comment]
Weird failure. Let me try to see if I can make it rerun, might be a fluke.

@ofborg eval ofborg-eval-check-meta

**Comment by dylanmtaylor** [comment]
@ofborg eval ofborg-eval-check-meta

**Comment by dylanmtaylor** [comment]
I just rebased onto latest master so that the suite of tests would rerun. I have a feeling it failed errantly on that one since I don't remember it failing on that before previously.

**Comment by dylanmtaylor** [comment]
Looks like that meta check worked fine on rerun and it is now building fine on Linux platforms in OfBorg :)

**Review by lilyinstarlight — APPROVED** [review]
Builds fine and seems to run fine for me now

**Comment by dylanmtaylor** [comment]
> Builds fine and seems to run fine for me now

Excellent! Can you please merge it in when the tests pass? I'm not sure why it's stuck on a darwin test if it's set to platform.linux.

**Comment by lilyinstarlight** [comment]
> Excellent! Can you please merge it in when the tests pass?

Yep I'm actually just giving @jtojnar a little bit of time if they had further input and otherwise planned to merge soon, don't worry :)

> I'm not sure why it's stuck on a darwin test if it's set to platform.linux.

Yeah the queue for aarch64-darwin is pretty long and it waits for a builder to be ready before it realizes eval will fail (which it should be able to tell from the path changes it calculates whether or not it even needs to queue darwin builds, but there may be reasons OfBorg doesn't do that and it might even be they just never implemented it)

**Review by jtojnar — APPROVED** [review]
The expression looks fine to me.

**Pull request merged**
